compiler/src/iree/compiler/Dialect/HAL/Transforms/MaterializeResourceCaches.cpp

```diff
@@ -20,322 +22,692 @@
+  Value initializerValue;
+};
+using DescriptorSetLayoutKey =
+    std::pair<ArrayAttr, IREE::HAL::DescriptorSetLayoutFlags>;
+
+struct PipelineLayout {
+  // All locations that use the layout.
+  SetVector<Location> locs;
+  // Lookup ops for this layout.
+  SmallVector<IREE::HAL::PipelineLayoutLookupOp> lookupOps;
+  // Global once materialized.
+  IREE::Util::GlobalOpInterface globalOp;
+  // Value within the initializer once materialized.
+  Value initializerValue;
+};
 
-    // Declare executable variables so that we can reference them during lookup
-    // replacement.
-    for (auto executableOp : executableOps) {
-      defineExecutableOp(executableOp);
-    }
+struct Executable {
+  // All locations that use the executable.
+  SetVector<Location> locs;
+  // Executable representing the program to load.
+  IREE::HAL::ExecutableOp executableOp;
+  // Lookup ops for this executable.
+  SmallVector<IREE::HAL::ExecutableLookupOp> lookupOps;
+  // Global once materialized.
+  IREE::Util::GlobalOpInterface globalOp;
+};
 
-    // Generate cached resource singletons and replace lookup ops with direct
-    // loads from variables.
-    for (auto lookupOp : pipelineLayoutLookupOps) {
-      replacePipelineLayoutLookupOp(lookupOp);
-    }
-    for (auto lookupOp : executableLookupOps) {
-      replaceExecutableLookupOp(lookupOp);
-    }
-  }
+struct DeviceResources {
+  DeviceResources() = default;
+  explicit DeviceResources(IREE::Util::GlobalOpInterface deviceOp)
+      : deviceOp(deviceOp) {}
+
+  // Global !hal.device.
+  IREE::Util::GlobalOpInterface deviceOp;
+
+  // Fallback devices that should be checked for resources.
+  // These are derived from the transitive set of #hal.device.fallback attrs.
+  SetVector<DeviceResources *> fallbackDeviceResources;
+
+  // Descriptor set layouts used on the device, keyed by [bindingAttrs, flags].
+  llvm::MapVector<DescriptorSetLayoutKey, DescriptorSetLayout>
+      descriptorSetLayouts;
+  // Pipeline layouts used on the device, keyed by layout attr.
+  llvm::MapVector<IREE::HAL::PipelineLayoutAttr, PipelineLayout>
+      pipelineLayouts;
+  // Executables used on the device, keyed by name.
+  llvm::MapVector<StringAttr, Executable> executables;
+};
 
-private:
-  IREE::Util::GlobalOp
-  defineDescriptorSetLayoutOp(Location loc, ArrayAttr bindingAttrs,
-                              IREE::HAL::DescriptorSetLayoutFlags flags) {
-    std::pair<Attribute, IREE::HAL::DescriptorSetLayoutFlags> key = {
-        bindingAttrs, flags};
-    auto existingIt = descriptorSetLayoutCache_.find(key);
-    if (existingIt != descriptorSetLayoutCache_.end()) {
-      return existingIt->second;
-    }
+static std::string getDeviceNamePrefix(IREE::Util::GlobalOpInterface deviceOp) {
+  StringRef deviceName = deviceOp.getGlobalName().getValue();
+  if (deviceName.starts_with("__")) {
+    // Already prefixed.
+    return deviceName.str();
+  }
+  return ("__" + deviceName).str();
```

Comment:
I can never remember if this is a legal way to prefix a StringRef with a literal. Assuage my paranoia and do something like the following?

std::string prefixed;
prefixed.reserve(deviceName.size() + 2);
prefixed.append("__");
prefixed.append(deviceName);

Reply:
any stringref/twine added to any other string-like thing produces a twine, then str() allocates an std::string for that - so technically safe here but wouldn't be if "__" was the result of a function returning an std::string

will change to make it clearer/more robust!